common/beerocks/bcl/source/beerocks_event_loop_impl.cpp

```diff
@@ -0,0 +1,215 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include <bcl/beerocks_event_loop_impl.h>
+
+#include <sys/epoll.h>
+#include <sys/timerfd.h>
+#include <unistd.h>
+
+#include <easylogging++.h>
+
+namespace beerocks {
+
+//////////////////////////////////////////////////////////////////////////////
+////////////////////////// Local module definitions //////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+// Maximal number of events to process in a single epoll_wait call
+static constexpr int MAX_POLL_EVENTS = 17;
+
+//////////////////////////////////////////////////////////////////////////////
+/////////////////////////////// Implementation ///////////////////////////////
+//////////////////////////////////////////////////////////////////////////////
+
+EventLoopImpl::EventLoopImpl(std::chrono::milliseconds timeout) : m_timeout(timeout)
+{
+    m_epoll_fd = epoll_create1(0);
+    LOG_IF(m_epoll_fd == -1, FATAL) << "Failed creating epoll: " << strerror(errno);
+}
+
+EventLoopImpl::~EventLoopImpl()
+{
+    // Delete all the file descriptors in the poll
+    LOG(DEBUG) << "Removing " << m_fd_to_event_handlers.size() << " FDs from the event loop";
+
+    while (!m_fd_to_event_handlers.empty()) {
+        int fd = m_fd_to_event_handlers.begin()->first;
+        EventLoopImpl::remove_handlers(fd);
+    }
+
+    // Close the poll fd
+    LOG_IF(close(m_epoll_fd) == -1, ERROR)
+        << "Failed closing epoll file descriptor: " << strerror(errno);
+}
+
+bool EventLoopImpl::register_handlers(int fd, const EventLoop::EventHandlers &handlers)
+{
+    if (-1 == fd) {
```

Comment:
I believe that code checkers would complain about assignment in `if` statement anyway.

Reply:
A [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions) places the constant portion of the expression on the left side of the conditional statement. Static analyzers shouldn't complain if you write conditions this way.

Reply:
The main reason to use Yoda assignment is to protect from code like `if (fd = -1)`, but such mistakes should be caugth by code analyzers anyway.
But if you prefer this idiom, I don't mind.